### DISCUSSION TITLE
Strongswan on Debian Bullseye Fails To Initiate IKEv2 Session

### DISCUSSION THREAD

**Discussion by blueaxions** [strongswan/strongswan#1621] [General]
Hello,

I have Strongswan 5.9.1-1 running on Debian Bullseye and is configured to initiate sessions with a Responder. The configuration works flawlessly when using PSK authentication but when I switched to this configuration https://docs.strongswan.org/docs/5.9/config/quickstart.html Roadwarrior Case, running 'swanctl --initiate --child vpnclient' fails with the following errors even though I have the cert for the Responder in the ~/x509 directory and the cert for the CA in the ~/x509ca directory. Tried it after adding the public key for the Responder to the ~/pubkey directory but it didn't work, Also tried after I added it to the ~/ecdsa directory and it failed as well.

On a side note this same configuration works without any issues on a Red Hat Enterprise Linux 8.6 VM. In fact, there I only had to place the cert for the root CA in the ~/x509ca folder on the Initiator side and was able to establish a tunnel.

[IKE] initiating IKE_SA vpnclient[2] to X.X.X.X
[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
[NET] sending packet: from X.X.X.X[500] to X.X.X.Y[500] (240 bytes)
[NET] received packet: from X.X.X.X[500] to Y.Y.Y.Y[500] (248 bytes)
[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519
[IKE] establishing CHILD_SA vpnclient{3}
[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr CPRQ(ADDR DNS) SA TSi TSr N(MOBIKE_SUP) N(NO_ADD_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
[NET] sending packet: from X.X.X.X[4500] to X.X.X.Y[4500] (272 bytes)
[NET] received packet: from X.X.X.X[4500] to Y.Y.Y.Y[4500] (256 bytes)
[ENC] parsed IKE_AUTH response 1 [ IDr AUTH EAP/REQ/MD5 ]
**[IKE] no trusted ECDSA public key found for '<Responder FQDN>'
[ENC] generating INFORMATIONAL request 2 [ N(AUTH_FAILED) ]**
[NET] sending packet: from Y.Y.Y.Y[4500] to X.X.X.X[4500] (80 bytes)
initiate failed: establishing CHILD_SA 'vpnclient' failed

My swanctl.conf:

swanctl {

    # Plugins to load in swanctl.
    # load =

    # VICI socket to connect to by default.
    # socket = unix://${piddir}/charon.vici

}

connections {
    vpnclient {
        dpd_delay = 30
        dpd_timeout = 90
        version = 1
        remote_addrs = X.X.X.X
        vips = 0.0.0.0
        local {
            auth = eap
            id = <INITIATOR_FQDN>
        }
        remote {
            auth = pubkey
            id = <RESPONDER_FQDN>
        }
    children {
           vpnclient {
                remote_ts = X.X.X.0/27

                updown = /usr/local/libexec/ipsec/_updown iptables
                esp_proposals = aes128gcm128-x25519
                start_action = trap
            }
         }
    version = 2
    proposals = aes128-sha256-x25519
    }
}

secrets {
    eap-vpnclient {
      id = <INITIATOR_FQDN>
      secret = "some_string"
   }
}

Thanks in advance!

-SR

**Comment by blueaxions** [comment]
On the Responder I see these messages in the log file:

Mar 24 15:02:51 vpn02 charon-systemd[10659]: generating key for CURVE_25519 failed
Mar 24 15:02:51 vpn02 charon-systemd[10659]: generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
.....
.....
.....
Mar 24 15:02:51 vpn02 charon-systemd[10659]: selected peer config 'vpnserver'
Mar 24 15:02:51 vpn02 charon-systemd[10659]: initiating EAP_MD5 method (id 0x3F)
Mar 24 15:02:51 vpn02 charon-systemd[10659]: peer supports MOBIKE
Mar 24 15:02:51 vpn02 charon-systemd[10659]: authentication of <RESPONDER_FQDN> (myself) with ECDSA_WITH_SHA384_DER successful
Mar 24 15:02:51 vpn02 charon-systemd[10659]: generating IKE_AUTH response 1 [ IDr AUTH EAP/REQ/MD5 ]
Mar 24 15:02:51 vpn02 charon-systemd[10659]: sending packet: from X.X.X.X[4500] to Y.Y.Y.Y[4500] (256 bytes)
Mar 24 15:02:51 vpn02 charon-systemd[10659]: received packet: from Y.Y.Y.Y[4500] to X.X.X.X[4500] (80 bytes)
Mar 24 15:02:51 vpn02 charon-systemd[10659]: parsed INFORMATIONAL request 2 [ N(AUTH_FAILED) ]

-SR

**Comment by misch42** [comment]
This log says that the other side could not authenticate.

Please send the log of the other side.

Michael.

**Comment by blueaxions** [comment]
Here's more logs from the Responder side:

Mar 24 15:02:51 vpn02 charon-systemd[10659]: received packet: from Y.Y.Y.Y[500] to X.X.X.X[500] (240 bytes)
Mar 24 15:02:51 vpn02 charon-systemd[10659]: parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
Mar 24 15:02:51 vpn02 charon-systemd[10659]: Y.Y.Y.Y is initiating an IKE_SA
Mar 24 15:02:51 vpn02 charon-systemd[10659]: selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519
Mar 24 15:02:51 vpn02 charon-systemd[10659]: generating key for CURVE_25519 failed
Mar 24 15:02:51 vpn02 charon-systemd[10659]: generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
Mar 24 15:02:51 vpn02 charon-systemd[10659]: sending packet: from X.X.X.X[500] to Y.Y.Y.Y[500] (248 bytes)
Mar 24 15:02:51 vpn02 charon-systemd[10659]: received packet: from Y.Y.Y.Y[4500] to X.X.X.X[4500] (272 bytes)
Mar 24 15:02:51 vpn02 charon-systemd[10659]: parsed IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr CPRQ(ADDR DNS) SA TSi TSr N(MOBIKE_SUP) N(NO_ADD_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
Mar 24 15:02:51 vpn02 charon-systemd[10659]: looking for peer configs matching X.X.X.X[<RESPONDER_FQDN>]...Y.Y.Y.Y[<INITIATOR_FQDN>]
Mar 24 15:02:51 vpn02 charon-systemd[10659]: selected peer config 'vpnserver'
Mar 24 15:02:51 vpn02 charon-systemd[10659]: initiating EAP_MD5 method (id 0x3F)
Mar 24 15:02:51 vpn02 charon-systemd[10659]: peer supports MOBIKE
Mar 24 15:02:51 vpn02 charon-systemd[10659]: authentication of '<RESPONDER_FQDN>' (myself) with ECDSA_WITH_SHA384_DER successful
Mar 24 15:02:51 vpn02 charon-systemd[10659]: generating IKE_AUTH response 1 [ IDr AUTH EAP/REQ/MD5 ]
Mar 24 15:02:51 vpn02 charon-systemd[10659]: sending packet: from X.X.X.X[4500] to Y.Y.Y.Y[4500] (256 bytes)
Mar 24 15:02:51 vpn02 charon-systemd[10659]: received packet: from Y.Y.Y.Y[4500] to X.X.X.X[4500] (80 bytes)
Mar 24 15:02:51 vpn02 charon-systemd[10659]: parsed INFORMATIONAL request 2 [ N(AUTH_FAILED) ]
Mar 24 15:02:51 vpn02 charon-systemd[10659]: generating INFORMATIONAL response 2 [ ]
Mar 24 15:02:51 vpn02 charon-systemd[10659]: sending packet: from X.X.X.X[4500] to Y.Y.Y.Y[4500] (80 bytes)
Mar 24 15:03:21 vpn02 charon-systemd[10659]: deleting half open IKE_SA with Y.Y.Y.Y after timeout

The swanctl.conf from Responder:

swanctl {
# Plugins to load in swanctl.
# load =

# VICI socket to connect to by default.
# socket = unix://${piddir}/charon.vici
}
connections {
vpnserver {
version = 1
local_addrs = X.X.X.X
proposals = aes192gcm16-aes128gcm16-prfsha256-ecp256-ecp521,aes192-sha256-modp3072,default
rekey_time = 0s
pools = primary-pool
fragmentation = yes
dpd_delay = 30s
dpd_timeout = 90s
local-1 {
auth = pubkey
certs = <RESPONDER_FQDN>.pem
id = <RESPONDER_FQDN>
}
remote-1 {
auth = eap-md5
}
children {
ike-psk {
local_ts = X.X.X.0/27
rekey_time = 0s
dpd_action = clear
esp_proposals = aes128gcm128-x25519
start_action = trap
hostaccess = yes
updown = /usr/libexec/strongswan/_updown iptables
}
}
version = 2
send_certreq = no
proposals = aes128-sha256-x25519
}
}
pools {
primary-pool {
addrs = Y.Y.Y.0/24
}
}
secrets {
eap-client1 {
id = <INITIATOR_FQDN1>
secret = "SOME_STRING"
}
eap-client2 {
id = <INITIATOR_FQDN2>
secret = "SOME_STRING"
}
}

-SR

**Comment by blueaxions** [comment]
Hi,

Even though the error is about public key, do you think that I need to put the respective private keys on both sides in the ~/private folder to fix this issue?

-SR

**Comment by blueaxions** [comment]
Hi,

Any help on this would be appreciated? Not sure why this configuration works without any issues between a CentOS Responder and a CentOS Initiator but fails between a CentOS Responder and a Debian Bullseye Initiator. Does FIPS mode need to be turned on in the Debian Bullseye Initiator because it is turned on the Responder?

-SR

**Comment by tobiasbrunner** [comment]
> In fact, there I only had to place the cert for the root CA in the ~/x509ca folder on the Initiator side and was able to establish a tunnel.

It seems like you haven't done that on the Debian host as it does not send a certificate request for that CA. Which in turn causes the responder to not send its certificate (which the initiator then obviously can't verify hence the error message you see). While you have set `send_certreq = no` on the responder, I don't see it in the initiator's config. So the only reason for it not to send the certificate request is a missing CA certificate.

Note that `~` refers to a user's home directory, not sure if that's what you meant because that's not usually where these directories/files are located.

**Comment by blueaxions** [comment]
I confirmed that the Initiator has the root CA cert in the /etc/swanctl/x509ca directory. I tried adding 'send_certreq = no' but it still fails with the same error. Here's the 'ls' output:

[root@vpn02 x509ca]# pwd
/etc/strongswan/swanctl/x509ca
[root@vpn02 x509ca]# ls -al
total 4
drwxr-xr-x.  2 root root  27 Mar 23 10:25 .
drwxr-xr-x. 16 root root 218 Mar 25 17:37 ..
-rw-------.  1 root root 974 Mar 23 10:25 ca_root.pem
[root@vpn02 x509ca]#


Do you think this has anything to do with FIPS mode not turned on the Initiator?

-SR

**Comment by blueaxions** [comment]
I've also tried using DER format for the CA cert file and no luck!

Are there any CA cert directory directive that needs to be set in the swanctl.conf file on the Initiator side?

-SR

**Comment by blueaxions** [comment]
Running 'swanctl --load-creds' fails with this:

swanctl --load-creds
loading '/etc/swanctl/x509ca/ca_root.pem' failed: parsing X509 certificate failed

-SR

**Reply by tobiasbrunner** [reply]
Then that file is either invalid or you are missing a required plugin. What format does it have? What key type? Can you parse it with `pki --print`.

**Comment by blueaxions** [comment]
Also, does it matter if the Strongswan versions between the Responder and Initiator don't match?

Strongswan versions:
Responder - 5.9.10
Initiator - 5.9.1-1

-SR

**Reply by tobiasbrunner** [reply]
No

**Comment by blueaxions** [comment]
Can you let me know the name of the plugin I could be missing?

Here's the output from "openssl x509 -in ./ca_root.pem -text -noout":

Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
        Signature Algorithm: ecdsa-with-SHA384
        Issuer: C = US, O = ACME Corporation, OU = ACME Certification Authority, CN = ROOT ECC CA
        Validity
            Not Before: Aug 31 16:26:13 2022 GMT
            Not After : Aug 31 16:26:13 2042 GMT
        Subject: C = US, O = ACME Corporation, OU = ACME Certification Authority, CN = ROOT ECC CA
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (384 bit)
                pub:
                    <key_value>
                ASN1 OID: secp384r1
                NIST CURVE: P-384
        X509v3 extensions:
            X509v3 Authority Key Identifier:
                keyid:<key_id>

            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Key Usage: critical
                Digital Signature, Non Repudiation, Certificate Sign, CRL Sign
            X509v3 Subject Key Identifier:
                <subject_key_id>
            Authority Information Access:
                OCSP - URI:http://ca.example.net:8080/ca/ocsp

    Signature Algorithm: ecdsa-with-SHA384
              <key_value>

-SR

**Reply by tobiasbrunner** [reply]
For ECDSA you need the _openssl_ plugin. On Debian that's provided by the `libstrongswan-standard-plugins` package.

**Comment by blueaxions** [comment]
The libstrongswan-standard-plugins package is already installed.

Interestingly, while I get a printable output when I ran the 'openssl x509' command against the ca_root.pem, but when I run the 'pki --print' command against it, it just hangs. Not sure why openssl works but not 'pki --print'.

-SR

**Reply by tobiasbrunner** [reply]
You might have called the program incorrectly. Unless you pass the certificate via `--in ca_root.pem`, it expects input on `stdin` (e.g. via `< ca_root.pem`) and will block if there is none.

**Comment by blueaxions** [comment]
Ok, here's the output from 'pki --print --in ./ca_root.pem':

  subject:  "C=US, O=ACME Corporation, OU=ACME Certification Authority, CN=ROOT ECC CA"
  issuer:   "C=US, O=ACME Corporation, OU=ACMECertification Authority, CN=ROOT ECC CA"
  validity:  not before Aug 31 09:26:13 2022, ok
             not after  Aug 31 09:26:13 2042, ok (expires in 7093 days)
  serial:    01
  flags:     CA CRLSign self-signed
  OCSP URIs: http://ca.example.net:8080/ca/ocsp
  authkeyId: <value>
  subjkeyId: <value>
  pubkey:    ECDSA 384 bits
  keyid:     <value>
  subjkey:   <value>

-SR

**Reply by tobiasbrunner** [reply]
Looks OK. So is this certificate loaded by swanctl and the daemon? Or do you still get the error about failing to parse the certificate? If swanctl is fine, check the daemon's log, see #196 for recommended log levels.

**Comment by blueaxions** [comment]
This same certificate loads fine without any issues on a Centos/RHEL 8 initiator, its only an issue on the initiator running Debian Bullseye. But I'll add the log snippet and post the log output from Debian Bullseye here on Monday, the guy who can get me the output is off for the weekend.

-SR

**Comment by blueaxions** [comment]
Hi,

Attaching the logs from both the Responder & Initiator side. One thing I noticed is it did not create a charon-debug.log file in /var/log. We added the following log snippet to swanctl.conf and restarted Strongswan:

filelog {
        # since 5.7.0 the path to the log file has to be specified in a separate setting if it contains dots,
        # use an arbitrary name without dots for the section instead of the one given here
        charon-debug-log {
                # this setting is required with 5.7.0 and newer if the path contains dots
                path = /var/log/charon_debug.log
                time_format = %a, %Y-%m-%d, %H:%M:%S
                default = 2
                net = 1
                enc = 1
                asn = 1
                job = 1
                ike_name = yes
                append = no
                flush_line = yes
        }
}


[rhel-strongswan-responder-log.txt](https://github.com/strongswan/strongswan/files/11140712/rhel-strongswan-responder-log.txt)
[debian-strongswan-initiator-log.txt](https://github.com/strongswan/strongswan/files/11140713/debian-strongswan-initiator-log.txt)

-SR

**Reply by tobiasbrunner** [reply]
> We added the following log snippet to swanctl.conf

Logging is [configured in strongswan.conf](https://docs.strongswan.org/docs/5.9/config/logging.html).

````
Apr  3 09:39:28 vpnclient01 charon-systemd[8635]: openssl FIPS mode(1) unavailable
Apr  3 09:39:28 vpnclient01 charon-systemd[8635]: plugin 'openssl': failed to load - openssl_plugin_create returned NULL 
````
You have to disable the `fips_mode` option for the _openssl_ plugin in strongswan.conf as it prevents the plugin from getting loaded. And without the plugin ECDSA can't be used so the certificate isn't loaded.

I see that you asked about FIPS mode previously, and the answer is yes, you have to disable this on Debian (Ubuntu has a FIPS-certified version of OpenSSL, Debian does not).

**Comment by blueaxions** [comment]
Right, once I disabled FIPS mode, the openssl plugin loaded without any issues and the Initiator established the tunnel successfully. It also has the log file in /var/log now. But on the Responder side, its still not creating the charon_debug.log file in /var/log, even after adding the log file snippet to /etc/strongswan/strongswan.conf file and restarting the service. Not sure if this has something to do with RHEL versus Debian but it would be nice to get logging working on CentOS/RHEL also.

-SR

**Reply by tobiasbrunner** [reply]
Could be SELinux preventing the daemon from creating that file (check the system log), in which case you have to find a different path. But maybe you don't need to log to a file, you can use the same log levels (if that's what you're interested in) when logging via syslog or to systemd's journal (see [the docs](https://docs.strongswan.org/docs/5.9/config/logging.html)).

**Comment by blueaxions** [comment]
Thanks for the assistance so far.
Regarding FIPS mode, you mentioned its not currently suppored in Debian, do you know if its supported in SUSE Linux either SLE or SLE Micro? Because our goal is to eventually migrate to SLE or SLE Micro and run strongSwan on SLE.

-SR

**Reply by tobiasbrunner** [reply]
Yes, looks like [SLES](https://documentation.suse.com/sles/15-SP1/html/SLES-all/cha-security-fips.html) and [SLE Micro](https://documentation.suse.com/sle-micro/5.3/html/SLE-Micro-all/cha-security-fips.html) both support a FIPS mode (although the latter is not certified yet it seems).
